pkg/interop/native/ledger/transaction_signer.go

```diff
@@ -0,0 +1,145 @@
+	Type WitnessConditionType
+	// Depends on the witness condition Type, its value can be asserted to the
+	// certain structure according to the following rule:
+	// WitnessBoolean -> ConditionBoolean
+	// WitnessNot ->  ConditionNot
+	// WitnessAnd ->  ConditionAnd
+	// WitnessOr -> ConditionOr
+	// WitnessScriptHash -> ConditionScriptHash
+	// WitnessGroup -> ConditionGroup
+	// WitnessCalledByContract -> ConditionCalledByContract
+	// WitnessCalledByGroup -> ConditionCalledByGroup
+	Value interface{}
+}
+
+// WitnessConditionWithType represents a single witness condition for rule-based
+// witness that has a type and doesn't have value. Currently, only CalledByEntry
+// witness condition satisfies this requirement.
+type WitnessConditionWithType struct {
+	Type WitnessConditionType
+}
+
+// WitnessConditionType represents the type of rule-based witness condition.
+type WitnessConditionType byte
+
+// Various witness condition types
+const (
+	// WitnessBoolean is a generic boolean condition.
+	WitnessBoolean WitnessConditionType = 0x00
+	// WitnessNot reverses another condition.
+	WitnessNot WitnessConditionType = 0x01
+	// WitnessAnd means that all conditions must be met.
+	WitnessAnd WitnessConditionType = 0x02
+	// WitnessOr means that any of conditions must be met.
+	WitnessOr WitnessConditionType = 0x03
+	// WitnessScriptHash matches executing contract's script hash.
+	WitnessScriptHash WitnessConditionType = 0x18
+	// WitnessGroup matches executing contract's group key.
+	WitnessGroup WitnessConditionType = 0x19
+	// WitnessCalledByEntry matches when current script is an entry script or is
+	// called by an entry script.
+	WitnessCalledByEntry WitnessConditionType = 0x20
+	// WitnessCalledByContract matches when current script is called by the
+	// specified contract.
+	WitnessCalledByContract WitnessConditionType = 0x28
+	// WitnessCalledByGroup matches when current script is called by contract
+	// belonging to the specified group.
+	WitnessCalledByGroup WitnessConditionType = 0x29
+)
+
+// Various witness conditions that WitnessRule condition can be cast to if it
+// has Value.
+type (
+	// ConditionBoolean is a boolean condition type.
+	ConditionBoolean bool
+	// ConditionNot inverses the meaning of contained condition.
+	ConditionNot struct {
+		// Condition can be cast either to WitnessConditionWithType or to
+		// WitnessConditionWithTypeAndValue.
+		Condition interface{}
+	}
+	// ConditionAnd is a set of conditions required to match. Each list item can
+	// be cast either to WitnessConditionWithType or to
+	// WitnessConditionWithTypeAndValue.
+	ConditionAnd []interface{}
+	// ConditionOr is a set of conditions one of which is required to match. Each
+	// list item can be cast either to WitnessConditionWithType or to
+	// WitnessConditionWithTypeAndValue.
+	ConditionOr []interface{}
+	// ConditionScriptHash is a condition matching executing script hash.
+	ConditionScriptHash interop.Hash160
```

Comment:
But I'm not so sure these types are useful, we can just declare that the value is `interop.Hash160`.